compiler/rustc_passes/messages.ftl

```diff
@@ -53,15 +53,55 @@ passes_both_ffi_const_and_pure =
     `#[ffi_const]` function cannot be `#[ffi_pure]`
 
 passes_break_inside_async_block =
-    `{$name}` inside of an `async` block
-    .label = cannot `{$name}` inside of an `async` block
+    `{$name}` inside an `async` block
+    .label = cannot `{$name}` inside an `async` block
     .async_block_label = enclosing `async` block
 
+passes_break_inside_async_closure =
+    `{$name}` inside an `async` closure
+    .label = cannot `{$name}` inside an `async` closure
+    .async_closure_label = enclosing `async` closure
+
+passes_break_inside_async_function =
+    `{$name}` inside an `async` function
+    .label = cannot `{$name}` inside an `async` function
+    .async_function_label = enclosing `async` function
+
+passes_break_inside_async_gen_block =
+    `{$name}` inside an `async gen` block
+    .label = cannot `{$name}` inside an `async gen` block
+    .async_gen_block_label = enclosing `async gen` block
+
+passes_break_inside_async_gen_closure =
+    `{$name}` inside an `async gen` closure
+    .label = cannot `{$name}` inside an `async gen` closure
+    .async_gen_closure_label = enclosing `async gen` closure
+
+passes_break_inside_async_gen_function =
+    `{$name}` inside an `async gen` function
+    .label = cannot `{$name}` inside an `async gen` function
+    .async_gen_function_label = enclosing `async gen` function
+
 passes_break_inside_closure =
-    `{$name}` inside of a closure
-    .label = cannot `{$name}` inside of a closure
+    `{$name}` inside a closure
+    .label = cannot `{$name}` inside a closure
     .closure_label = enclosing closure
 
+passes_break_inside_gen_block =
+    `{$name}` inside a `gen` block
+    .label = cannot `{$name}` inside a `gen` block
+    .gen_block_label = enclosing `gen` block
+
+passes_break_inside_gen_closure =
+    `{$name}` inside a `gen` closure
+    .label = cannot `{$name}` inside a `gen` closure
+    .gen_closure_label = enclosing `gen` closure
+
+passes_break_inside_gen_function =
+    `{$name}` inside a `gen` function
+    .label = cannot `{$name}` inside a `gen` function
+    .gen_function_label = enclosing `gen` function
+
```

Comment:
You can combine this into one:

```
passes_break_inside_coroutine =
    `{$name}` inside an `{$coroutine_kind}` {$closure_kind}
    .label = cannot `{$name}` inside an `{$coroutine_kind}` {$closure_kind}
    .async_closure_label = enclosing `{$coroutine_kind}` {closure_kind}
```

Then you can set `coroutine_kind`  and `closure_kind` in the diagnostic.

Reply:
Then you can match on https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.CoroutineDesugaring.html and https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.CoroutineSource.html to populate these.

Reply:
Thanks for the suggestion.